tests/firedrake/regression/test_bddc.py

```diff
@@ -85,46 +97,90 @@ def solve_riesz_map(mesh, family, degree, bcs, condense):
+                                     solver_parameters=sp)
     solver.solve()
+    uerr = Function(V).assign(uh - u_exact)
+    assert (assemble(a(uerr, uerr)) / assemble(a(u_exact, u_exact))) ** 0.5 < 1E-8
     return solver.snes.getLinearSolveIterations()
 
 
 @pytest.fixture(params=(2, 3), ids=("square", "cube"))
-def mesh(request):
+def mh(request):
     dim = request.param
-    nx = 4
-    msh = UnitSquareMesh(nx, nx, quadrilateral=True)
+    nx = 3
+    base = UnitSquareMesh(nx, nx, quadrilateral=True)
+    mh = MeshHierarchy(base, 1)
     if dim == 3:
-        msh = ExtrudedMesh(msh, nx)
-    return msh
+        mh = ExtrudedMeshHierarchy(mh, height=1, base_layer=nx)
+    return mh
+
+
+@pytest.mark.parallel
+@pytest.mark.parametrize("degree", range(1, 3))
+@pytest.mark.parametrize("variant", ("spectral", "fdm"))
+def test_vertex_dofs(mh, variant, degree):
+    """Check that we extract the right number of vertex dofs from a high order Lagrange space."""
+    from firedrake.preconditioners.bddc import get_restricted_dofs
+    mesh = mh[-1]
+    P1 = FunctionSpace(mesh, "Lagrange", 1, variant=variant)
+    V0 = FunctionSpace(mesh, "Lagrange", degree, variant=variant)
+    v = get_restricted_dofs(V0, "vertex")
+    assert v.getSizes() == P1.dof_dset.layout_vec.getSizes()
 
 
 @pytest.mark.parallel
-@pytest.mark.parametrize("family", "Q")
-@pytest.mark.parametrize("degree", (4,))
+@pytest.mark.parametrize("family,degree", [("Q", 4)])
 @pytest.mark.parametrize("condense", (False, True))
-def test_bddc_fdm(mesh, family, degree, condense):
+def test_bddc_fdm(rg, mh, family, degree, condense):
+    """Test h-independence of condition number by measuring iteration counts"""
+    variant = "fdm"
     bcs = True
-    tdim = mesh.topological_dimension
-    expected = 7 if tdim == 2 else 11
-    assert solve_riesz_map(mesh, family, degree, bcs, condense) <= expected
+    its = [solve_riesz_map(rg, m, family, degree, variant, bcs, condense=condense) for m in mh]
+    assert (np.diff(its) <= 2).all()
+
+
+@pytest.mark.parallel
+@pytest.mark.parametrize("family,degree", [("Q", 4)])
+@pytest.mark.parametrize("vector", (False, True), ids=("scalar", "vector"))
+def test_bddc_aij_quad(rg, mh, family, degree, vector):
+    """Test h-independence of condition number by measuring iteration counts"""
+    variant = None
+    bcs = True
+    its = [solve_riesz_map(rg, m, family, degree, variant, bcs, vector=vector) for m in mh]
+    assert (np.diff(its) <= 1).all()
+
+
+@pytest.mark.parallel
+@pytest.mark.parametrize("family,degree", [("CG", 3), ("N1curl", 3), ("N1div", 3)])
+def test_bddc_aij_simplex(rg, family, degree):
+    """Test h-independence of condition number by measuring iteration counts"""
+    variant = None
+    bcs = True
+    meshes = [UnitCubeMesh(nx, nx, nx) for nx in (3, 6)]
+    its = [solve_riesz_map(rg, m, family, degree, variant, bcs) for m in meshes]
+    assert (np.diff(its) <= 1).all()
```

Comment:
@JHopeCollins here we are now testing theoritical bounds, but we might be in a pre-assymptotic regime still. I wouldn't say this is completely robust with respect to stochastic things like mesh partitioning. Iterations might suddenly differ by 2 instead of 1. 

I think we can compute the condition number of the preconditioned system projected into the Krylov space.